src/internal/ReplaySubject.ts

```diff
@@ -70,15 +70,36 @@ export class ReplaySubject<T> extends Subject<T> {
     this._throwIfClosed();
     this._trimBuffer();
 
-    const subscription = this._innerSubscribe(subscriber);
-
     const { _infiniteTimeWindow, _buffer } = this;
-    // We use a copy here, so reentrant code does not mutate our array while we're
-    // emitting it to a new subscriber.
-    const copy = _buffer.slice();
-    for (let i = 0; i < copy.length && !subscriber.closed; i += _infiniteTimeWindow ? 1 : 2) {
-      subscriber.next(copy[i] as T);
+
+    // Replay the current values, but also the ones that are re-entered synchronously while emitting them.
+    const syncBuffer = _buffer.slice();
+    let isInitializing = true;
+
+    // This subscriber prevents new values from being emitted before the current values are all flushed.
+    const syncSubscriber = new SafeSubscriber<T>({
+      next(value) {
+        if (isInitializing) {
+          syncBuffer.push(value);
+        } else {
+          subscriber.next(value);
+        }
+      },
+      error(e) {
+        subscriber.error(e);
+      },
+      complete() {
+        subscriber.complete();
+      },
```

Comment:
That's arguable though, it's what I mentioned on the description of this PR.

Note that this `syncSubscriber` is just mirroring the original `subscriber`. What I mean is that if previously it was receiving error / complete, then now it's also receiving it. I decided not to change it now because I think it's up for discussion.

As an example how forwarding error would be inconsistent with other operators is `delay`. `delay` propagates the error immediately even if there are still values on the way:

```ts
const obs$ = concat(
  [1],
  [2],
  throwError(() => new Error("hey"))
).pipe(
  delay(100)
);

obs$.subscribe({
  next(v) { // never called
    console.log(v);
  },
  error(e){ // Immediately called
    console.error(e);
  }
})
```

I can't find any example for `complete` though, so that's probably something to change too.

Reply:
the contract for ReplaySubject is different here:
```ts
const obs$ = concat(
  [1],
  [2],
  throwError(() => new Error("hey"))
).pipe(
  share({
    connector: () => new ReplaySubject(),
    resetOnError: false,
    resetOnRefCountZero: false,
  }),
);

obs$.subscribe({
  next(v) {
    console.log(v);
  },
  error(e){
    console.error(e);
  }
})
// log 1
// log 2
// error Error(hey)
```

Reply:
This will still work with the implementation of this PR - This is not testing anything about having errors entering while the initial buffer is being emitted.

Handling re-entrant values/errors/completes while the initial subscription is happening is something that's currently undefined behaviour - And I wondered whether the original PR would be closed instead as such (and probably also my implementation with it).

Really, I can also add in this change for error/complete, but I'd like to get confirmation from someone on the core RxJS team that this case should also be handled.

Reply:
I pushed it as a separate commit so it can be easily reverted if needed.